services/paymaster/src/rpc.rs

```diff
@@ -153,41 +156,111 @@ impl ChainIndex {
+                tracing::Span::current().record("result", "preflight_failure");
+                tracing::warn!(
+                    "Transaction {} failed preflight: {}",
+                    transaction.signatures[0],
+                    error
+                );
                 return Ok(ConfirmationResultInternal::UnconfirmedPreflightFailure {
                     signature: transaction.signatures[0],
                     error,
                 });
             }
+            tracing::Span::current().record("result", "send_failed");
+            tracing::error!(
+                "Failed to send transaction {}: {}",
+                transaction.signatures[0],
+                err
+            );
             return Err(to_error_response(err));
         }
     };
 
-    confirm_transaction(pubsub, signature, Some(rpc.commitment())).await
+    confirm_transaction_with_reconnect(pubsub, signature, Some(rpc.commitment())).await
 }
 
 pub const CONFIRMATION_TIMEOUT: Duration = Duration::from_secs(60);
 
 #[tracing::instrument(
     skip_all,
-    fields(tx_hash = %signature)
+    name = "confirm_transaction",
+    fields(
+        tx_hash = %signature,
+        result
+    )
 )]
+async fn confirm_transaction_with_reconnect(
+    pubsub: &PubsubClientWithReconnect,
+    signature: Signature,
+    commitment: Option<CommitmentConfig>,
+) -> Result<ConfirmationResultInternal, ErrorResponse> {
+    let result = {
+        let pubsub_guard = pubsub.client.load();
+        confirm_transaction(&pubsub_guard, signature, commitment).await
+    };
+
+    match result {
+        Ok(confirmation) => {
+            tracing::Span::current().record("result", confirmation.status_string().as_str());
+            Ok(confirmation)
+        }
+        Err(e) if is_subscription_error(e.0, &e.1) => {
+            tracing::warn!("WebSocket subscription failed, attempting reconnection and retry");
+            pubsub.reconnect_pubsub().await?;
+            let new_guard = pubsub.client.load();
+            let retry_result = confirm_transaction(&new_guard, signature, commitment)
+                .await
+                .map_err(|e| e.into());
+
+            match retry_result {
+                Ok(ref confirmation) => {
+                    tracing::Span::current()
+                        .record("result", confirmation.status_string().as_str());
+                }
+                Err(_) => {
+                    tracing::Span::current().record("result", "unconfirmed");
+                }
+            }
+
+            retry_result
+        }
+        Err(e) => {
+            tracing::Span::current().record("result", "unconfirmed");
+            Err(e.into())
+        }
+    }
+}
+
+pub const SIGNATURE_SUBSCRIPTION_ERROR: &str = "Failed to subscribe to signature";
+
+fn is_subscription_error(status_code: StatusCode, error_string: &str) -> bool {
+    status_code == StatusCode::BAD_GATEWAY && error_string == SIGNATURE_SUBSCRIPTION_ERROR
```

Comment:
i don't love the string matching here, but i didn't want to change the error status code or rearchitect the error types too much just for this